lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVM.cpp

```diff
@@ -2661,75 +2659,207 @@ struct DotOpConversionHelper {
+    matShapeK = matShape[2];
+
+    int NK = aShape[1];
+    // shape / shape_per_cta
+    numRepM = std::max<int>(dShape[0] / (wpt[0] * mmaInstrM), 1);
+    numRepN = std::max<int>(dShape[1] / (wpt[1] * mmaInstrN), 1);
+    numRepK = std::max<int>(NK / mmaInstrK, 1);
+
+    Value _32 = i32_val(32);
+    lane = urem(thread, _32);
+    warp = udiv(thread, _32);
+    warpMN = udiv(warp, i32_val(wpt[0]));
+    warpM = urem(warp, i32_val(wpt[0]));
+    warpN = urem(warpMN, i32_val(wpt[1]));
+
+    aElemBytes = aTensorTy.getElementTypeBitWidth() / 8;
+    bElemBytes = bTensorTy.getElementTypeBitWidth() / 8;
+  }
+
+  // Loading $a from smem to registers, returns a LLVM::Struct.
+  Value loadA() {
+    ValueTable ha;
+    std::function<void(int, int)> loadFn;
+    if (aTensorTy.getEncoding().isa<SharedEncodingAttr>()) {
+      // load from smem
+      loadFn = getLoadMatrixFn(
+          A, adapter.a() /*llTensor*/, mmaLayout.getWarpsPerCTA()[0] /*wpt*/,
+          1 /*kOrder*/, {mmaInstrM, mmaInstrK} /*instrShpae*/,
+          {matShapeM, matShapeK} /*matShape*/, warpM /*warpId*/, ha /*vals*/);
+    } else if (aTensorTy.getEncoding().isa<BlockedEncodingAttr>()) {
+      // load from registers, used in gemm fuse
+      // TODO(Superjomn) Port the logic.
+      assert(false && "Loading A from register is not supported yet.");
+    } else {
+      assert(false && "A's layout is not supported.");
+    }
 
-  auto wpt = mmaLayout.getWarpsPerCTA();
+    // step1. Preform loading.
+    for (unsigned m = 0; m < numRepM; ++m)
+      for (unsigned k = 0; k < numRepK; ++k)
+        loadFn(2 * m, 2 * k);
 
-  // TODO(Superjomn) Process C->is_trans_a() logic
+    // step2. Format the values to LLVM::Struct to passing to mma codegen.
+    Value result = composeValuesToDotOperandLayoutStruct(ha, numRepM, numRepK);
 
-  DotOpConversionHelper helper(op);
+    // TODO[Superjomn]: Replace the convert_layout op with the result once the
+    // DotOperandEncodingAttr is ready.
+    return result;
+  }
 
-  int NK = aShape[1];
+  // Loading $b from smem to registers, returns a LLVM::Struct.
+  Value loadB() {
+    ValueTable hb;
+    auto loadFn = getLoadMatrixFn(
+        B, adapter.b() /*llTensor*/, mmaLayout.getWarpsPerCTA()[1] /*wpt*/,
+        0 /*kOrder*/, {mmaInstrK, mmaInstrN} /*instrShpae*/,
+        {matShapeK, matShapeN} /*matShape*/, warpN /*warpId*/, hb /*vals*/);
 
-  auto mmaInstrShape = helper.getMmaInstrShape();
-  const int mmaInstrM = mmaInstrShape[0];
-  const int mmaInstrN = mmaInstrShape[1];
-  const int mmaInstrK = mmaInstrShape[2];
+    for (unsigned n = 0; n < std::max(numRepN / 2, 1); ++n) {
+      for (unsigned k = 0; k < numRepK; ++k)
+        loadFn(2 * n, 2 * k);
+    }
 
-  auto matShape = helper.getMmaMatShape();
-  const int matShapeM = matShape[0];
-  const int matShapeN = matShape[1];
-  const int matShapeK = matShape[2];
+    Value result = composeValuesToDotOperandLayoutStruct(
+        hb, std::max(numRepN / 2, 1), numRepK);
```

Comment:
Why here `numRepN / 2`？

Reply:
Because the loading is for `ldmatrix.16816`, each `ldmatrix.16816` needs two `ldmatrix.m8n8` to load two 8x8 matrix from B, the N-axis loads just half numbers compared to K-axis.

```
_____K___
| x x | x x | \ N
| x x | x x | / 
```

That's why the loading for-loop also loop `numRepN/2` times
```c++
for (unsigned n = 0; n < std::max(numRepN / 2, 1); ++n) {
      for (unsigned k = 0; k < numRepK; ++k)
        loadFn(2 * n, 2 * k); // load 2x2 elements each time
    }
```